fecgov/fec-cms
# Feature/2682 new js testing pkg

# = = DO NOT MERGE = =

## Summary
Lots of changes here—not a lot is touched but all front-end (that's tested) is possibly affected.

The previous ticket, [2681](https://github.com/fecgov/fec-cms/issues/2681), was to research various JavaScript testing options. This PR tests
- swapping out PhantomJS for [Puppeteer](https://github.com/GoogleChrome/puppeteer)
- upgrading various Node packages—only as far as needed

Additionally contains
- various linting changes
- some transitions to ES6 (Underscore to Object.assign(), var to const, etc)
- a fix for a test at line 322 of fec/fec/tests/js/top-entities-breakdown.js

Resolves # [2682](https://github.com/fecgov/fec-cms/issues/2682)

## Impacted areas of the application
- package.json was affected for the upgrades.
- removed a failing test in fec/fec/tests/js/filter-panel.js
- fixed a test in fec/fec/tests/js/top-entities-breakdown.js
- ES5 > ES6 changes shouldn't impact the app now that we've upgraded
- linting fixes shouldn't impact the application

**Note:** This ticket specifically *does not* upgrade [webpack](https://webpack.js.org) or its dependencies. There's a collision between a webpack sub package that doesn't like how we're building our chunks. (Upgrading to webpack 4+ will require overhauling how we're bundling our js files)


## Screenshots
No screenshots—it's all code.

## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
develop test/2653-upgrade-node | [2688](https://github.com/fecgov/fec-cms/pull/2688)

## How to test
Just like [2688](https://github.com/fecgov/fec-cms/pull/2688),
- `nvm ls` will list which versions of Node are installed locally. `node -v` will show the current version. This PR is using the (current) latest LTS, 10.15.1. To install it, use `nvm install v10.15.1`. Switch to it with `nvm use 10.15.1`
- `npm install -g npm` will update npm to the latest (currently 6.8.0)
* It would be a good idea to remove the node_modules directory, either delete it through Finder or use rm -rf node_modules/`
- After removing node_modules, clear the cache with `npm cache clean --force`
- `npm i` will go get versions of everything in package.json
- Will probably need to run `npm rebuild node-sass`
- `npm run build` (inside the fec-cms pyenv instance) will run webpack and build-sass
- It may be necessary to update the linter used by the code editor/IDE. For Visual Studio Code, I had to deactivate/remove jslint and jshint
- Run the server and/or watch as usual
- Running `npm run test-single` will run `test-single` from the scripts section of package.json, which will run the linter and start Karma using ChromeHeadless.

The tests (lint and Karma) should run fine and the site should behave as it normally does.

____
